src/Microsoft.ML.Data/Evaluators/RankingEvaluator.cs

```diff
@@ -946,23 +948,34 @@ private protected override IEnumerable<string> GetPerInstanceColumnsToSave(RoleM
 
     internal static class RankingUtils
     {
+        private static readonly object _lock = new object();
         private static volatile Double[] _discountMap;
+        private static volatile int _maxTruncationLevel = 0;
+
         public static Double[] DiscountMap
         {
             get
             {
-                double[] result = _discountMap;
-                if (result == null)
+                return _discountMap;
+            }
+        }
+        /// <summary>
+        /// Reallocates discountMap for the largest truncationLevel seen so far
+        /// </summary>
+        public static void SetTruncationLevel(int truncationLevel)
```

Comment:
In general I find this pattern not following good practices. Maintaining mutable global state is hard to do right.

How large do we expect this `Double[] DiscountMap` to get in general? If it was limited to `100` before, maybe we could change this code a bit so we don't need to mutate shared state.

Instead, what if we had a method to get the `Double[] DiscountMap` like:

```C#
static double[] GetDiscountMap(int truncationLevel)
```

And inside that method we check if `truncationLevel <= some threshold (like 100)`, and if it is less than 100, then we return a cached, read-only `double[]` that was precomputed. That way most normal usages will get the cached array and won't allocate new memory. If  `truncationLevel` is bigger, than we allocate a new array, fill it up, and return it. But we don't cache it.

Thoughts?

Reply:
Let us say, two different callers set two different truncations levels of N and 2N. In the model you suggest, we will return two different arrays for each of them. But the first N entries of both those arrays would be identical and duplicated space. (In theory it is an actual savings, but it is just a bunch of floats even when the truncation limit is high. So the practical impact is minimal.)

When complex thread interactions and state is involved, I agree maintaining mutable global state is difficult. But this is a fairly simple case that is protected by a lock.  (So, in theory, it is a complex task, but right here this seems straightforward)

As I was writing this, I realized there is another approach. The value of the discount map entry is completely determined by the index of the entry (1 / Math.Log(2 + i)). So we don't need to maintain a statically computed map at all, and simply compute it dynamically each time it is accessed. We would be trading off a small gain in memory consumption for a small loss in runtime performance during evaluation. 

If we wanted to optimize that even further, we would try to maintain a sparse array where the value is calculated dynamically and cached.

The overall gains in any approach would be minimal. I think we have hit the point of diminishing returns and it is okay to leave it as is.


---
In reply to: [333065450](https://github.com/dotnet/machinelearning/pull/4081#discussion_r333065450) [](ancestors = 333065450)

Reply:
> The overall gains in any approach would be minimal. 

I totally disagree with this statement. Following good programming practices is not a "minimal" gain IMO.

> Let us say, two different callers set two different truncations levels of N and 2N. In the model you suggest, we will return two different arrays for each of them.

This is only true when `2N > 100`, which wasn't even supported before this change. So the cases that actually would be doing this will not be the majority case.

Reply:
100 was an arbitrarily chosen number. That argument wouldn't have been true if the previous chosen value was 10 instead of 100. Also the very idea of this fix is to support truncation levels > 100 and enable such cases in the future. 

I don't think of good programming practice as an absolute definition. There are three approaches here and there are tradeoffs with each approach.

Approach 1. Use a static class and protect state correctly with a lock.
- Pros: Uses the minimum memory and computation necessary 
- Cons: Has mutable global state (but is protected with a lock)

Approach 2: Create a new discount map for each caller that asks for a higher truncation level than current
- Pros: Eliminates mutable state
- Cons: Consumes more memory than necessary. The caller can in theory optimize this by determining the highest truncation level they will need in the lifetime of the process and set that truncation level first. But that would require the caller to know implementation specifics. 

Approach 3: Compute discount map value based on index dynamically
- Pros: No global state and consumes the least memory possible
- Cons: A minor calculation is done repeatedly during evaluation

As I mentioned above, I think we have hit the point of diminishing returns and am happy to leave it as is. But I am also looking to make progress on this PR. Since you feel strongly about this, we will make the change per your suggestion. 



---
In reply to: [333204539](https://github.com/dotnet/machinelearning/pull/4081#discussion_r333204539) [](ancestors = 333204539)

Reply:
I had another question. You proposed caching a precomputed array for a threshold of say 100. That once again special cases 100 for truncation level and the whole point of this fix was to remove that. Since you are okay with returning new arrays for all values greater than 100, why not return a new array for 100 too. 
Wouldn't that be a simpler implementation?

---
In reply to: [333234894](https://github.com/dotnet/machinelearning/pull/4081#discussion_r333234894) [](ancestors = 333234894,333204539)

Reply:
It would be simpler, yes. But the reason I proposed it is to address your cons with Approach 2 above - `Consumes more memory than necessary`.  In my proposed approach, the caller will never need to care. They get the `truncationLevel` passed in through the Arguments class, and they just call `GetDiscountMap(truncationLevel)` when they need the DiscountMap. Inside of the `GetDiscountMap` it checks if the cached, pre-computed array is large enough to handle the request. If it is large enough (i.e. `truncationLevel <= 100`) it returns the cached array. If it is not large enough (`truncationLevel > 100`, which will barely ever happen), then it calculates a new array and hands it out.

> Since you are okay with returning new arrays for all values greater than 100, why not return a new array for 100 too.

The reason is because in the majority of the use cases, `100` will work. Even in the issue being addressed #3993, it says:

> While I'm unsure what the MaxTruncationLevel value should be, I have seen on a ranking contest\example on Kaggle.com where one contest was measuring NDCG with a truncation level of up to 38.

And `38` is a lot less than `100`.